### PR TITLE
ModelHub BQ: Add conversion event

### PR DESCRIPTION
**Pull request by KathiaBarahona**
Probably will close this, this is just a very nasty way to slice with dicts for BQ json. This just proves that most things tested in `test_modelhub.py` actually work (just slicing problem, document with proposal https://docs.google.com/document/d/1yPt4ML5liYgqiRbbJmhW_j3Co5PkSA0539nKocb25xg/edit#)

Edit: removed the temporal json slicing from tests. 
Opened a PR for support of slicing in bach: https://github.com/objectiv/objectiv-analytics/pull/852

### CODE REVIEW

File: modelhub/modelhub/map.py [hunk 1]
```diff
@@ -58,31 +58,31 @@ def is_new_user(self, data: bach.DataFrame, time_aggregation: str = None) -> bac
...
+        if window.base_node != window_ta.base_node:
+            data_cp = data_cp.materialize(node_name='time_agg_window')
+
+        session_id_series = data_cp['session_id']
+        is_first_session = session_id_series.min(partition=window)
+        is_first_session_time_aggregation = session_id_series.min(partition=window_ta)
+
+        is_new_user_series = is_first_session_time_aggregation == is_first_session
+        is_new_user_series = is_new_user_series.copy_override_type(bach.SeriesBoolean)
+        return is_new_user_series.copy_override(name='is_new_user').materialize()
```
Comment (KathiaBarahona) [inline modelhub/modelhub/map.py:85]: Done here https://github.com/objectiv/objectiv-analytics/pull/800

File: modelhub/modelhub/map.py [hunk 2]
```diff
@@ -197,22 +198,28 @@ def pre_conversion_hit_number(self,
...
-
-        pre_conversion_hits = pre_conversion_hits.materialize()
-        data['pre_conversion_hit_number'] = pre_conversion_hits['pre_conversion_hit_number']
-
-        return data.pre_conversion_hit_number
+        # make the query more clean, just require these series
+        required_series = [partition, 'session_id', 'session_hit_number', '__conversions']
+        data = data[required_series]
+
+        window = data.groupby(partition).window(start_boundary=None, end_boundary=None)
```
Comment (KathiaBarahona) [inline modelhub/modelhub/map.py:205]: had to remove boundaries, BQ was generating errors

Comment (KathiaBarahona) [reply modelhub/modelhub/map.py]: You were right, we just need to sort the window function. Done here https://github.com/objectiv/objectiv-analytics/pull/838/commits/857dd41daf018eb43b81a6050fb4d150132ab486

**Review by vard-obj — APPROVED** [review]
lgtm

File: modelhub/modelhub/map.py [hunk 3]
```diff
@@ -197,22 +198,29 @@ def pre_conversion_hit_number(self,
...
+
+        # sort all windows by session id and hit number
+        sort_by = ['session_id', 'session_hit_number']
 
-        data['__is_converted'] = converted != 0
-        data = data.materialize()
-        assert(isinstance(data['__is_converted'], SeriesBoolean))  # help mypy to overcome generic Series type
-        pre_conversion_hits = data[data['__is_converted']]
-        pre_conversion_hits = pre_conversion_hits[pre_conversion_hits['__conversions'] == 0]
+        window = data.sort_values(sort_by, ascending=False).groupby(partition).window()
```
Comment (jansentom) [inline modelhub/modelhub/map.py:208]: Shouldn't this be sorted ascending, else the last conversion in the session comes first?

Comment (hendrik-obj) [reply modelhub/modelhub/map.py:208]: i think it is correct, as it uses conversions_in_time, that one has 0 for all hits _before_ conversion. We want those also to have `__is_converted` to be `True`

**Review by hendrik-obj — COMMENTED** [review]
lgtm, but maybe we can merge it to main (with just the changes to these two files)? Because it also works on main, and might be easier to review the kathiabarahona/bq_json_slicing_with_dicts branch

**Review by jansentom — APPROVED** [review]

**Pull request merged**